older versions of libsystemd lack all functions use for our `bus` module

In particular, ubuntu-18.04's libsystemd lacks 2 functions which result in link errors (when the `bus` feature is not disabled):

           /home/runner/work/rust-systemd/rust-systemd/src/bus/mod.rs:1269: undefined reference to `sd_bus_slot_set_destroy_callback'
          /home/runner/work/rust-systemd/rust-systemd/src/bus/mod.rs:1270: undefined reference to `sd_bus_slot_set_floating'


We should consider looking into whether these can be avoided an provide a usable API in `systemd` (high level crate). Possible that `bus` (the feature) could be subsetted (so some of bus is usable with old `libsystemd`. Minimally, detecting the presence of these functions at configuration time and avoiding emitting the rust definitions would help us get better error
messages (ie: at compile instead of at link).

openssl has to deal with similar issues and probably has some approach we can adapt.

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -13,8 +13,15 @@ edition = "2018"
 [features]
 default = ["bus", "journal"]
 
+full = [
+  "bus",
+  "journal",
+  "systemd_v245"
+]
+
 bus = ["libsystemd-sys/bus"]
 journal = ["libsystemd-sys/journal"]
+systemd_v245 = ["libsystemd-sys/systemd_v245"]
 
 [dependencies]
 log = "~0.4"
@@ -37,3 +44,7 @@ doc-comment = "0.3"
 
 [profile.release]
 debug = true
+
+[package.metadata.docs.rs]
+all-features = true
+rustdoc-args = ["--cfg", "docsrs"]

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -13,12 +13,6 @@ edition = "2018"
 [features]
 default = ["bus", "journal"]
 
-full = [
-  "bus",
-  "journal",
-  "systemd_v245"
-]
-
 bus = ["libsystemd-sys/bus"]
 journal = ["libsystemd-sys/journal"]
 systemd_v245 = ["libsystemd-sys/systemd_v245"]

--- a/libsystemd-sys/Cargo.toml
+++ b/libsystemd-sys/Cargo.toml
@@ -17,6 +17,7 @@ default = ["bus", "journal"]
 
 bus = []
 journal = []
+systemd_v245 = []
 
 [dependencies]
 libc = "0.2.76"

--- a/libsystemd-sys/src/journal.rs
+++ b/libsystemd-sys/src/journal.rs
@@ -26,6 +26,7 @@ extern "C" {
     // (we don't need to do c-style format strings)
 
     pub fn sd_journal_open(ret: *mut *mut sd_journal, flags: c_int) -> c_int;
+    #[cfg(feature = "systemd_v245")]
     pub fn sd_journal_open_namespace(
         ret: *mut *mut sd_journal,
         namespace: *const c_char,

--- a/src/journal.rs
+++ b/src/journal.rs
@@ -408,7 +408,7 @@ impl OpenOptions {
     ///
     /// `sd_journal_open()`: https://www.freedesktop.org/software/systemd/man/sd_journal_open.html
     pub fn open(&self) -> Result<Journal> {
-        Journal::open_with_opts_ns::<&std::ffi::CStr>(None, self)
+        Journal::open_with_opts::<&std::ffi::CStr>(self)
     }
 
     /// Open the log journal for reading in the given namespace. Entries included are dependent on
@@ -422,6 +422,8 @@ impl OpenOptions {
     /// This corresponds to [`sd_journal_open_namespace()`]
     ///
     /// `sd_journal_open_namespace()`: https://www.freedesktop.org/software/systemd/man/sd_journal_open.html
+    #[cfg(feature = "systemd_v245")]
+    #[cfg_attr(docsrs, doc(cfg(feature = "systemd_v245")))]
     pub fn open_namespace<A: CStrArgument>(&self, namespace: A) -> Result<Journal> {
         Journal::open_with_opts_ns(Some(namespace), self)
     }
@@ -520,6 +522,36 @@ impl OpenFilesOptions {
 }
 
 impl Journal {
+    fn open_with_opts<A: CStrArgument>(opts: &OpenOptions) -> Result<Journal> {
+        let mut flags = opts.extra_raw_flags;
+        if opts.current_user {
+            flags |= ffi::SD_JOURNAL_CURRENT_USER;
+        }
+        if opts.system {
+            flags |= ffi::SD_JOURNAL_SYSTEM;
+        }
+        if opts.local_only {
+            flags |= ffi::SD_JOURNAL_LOCAL_ONLY;
+        }
+        if opts.local_only {
+            flags |= ffi::SD_JOURNAL_LOCAL_ONLY;
+        }
+        if opts.runtime_only {
+            flags |= ffi::SD_JOURNAL_RUNTIME_ONLY;
+        }
+        if opts.all_namespaces {
+            flags |= ffi::SD_JOURNAL_ALL_NAMESPACES;
+        }
+        if opts.include_default_namespace {
+            flags |= ffi::SD_JOURNAL_INCLUDE_DEFAULT_NAMESPACE;
+        }
+
+        let mut jp = MaybeUninit::uninit();
+        crate::ffi_result(unsafe { ffi::sd_journal_open(jp.as_mut_ptr(), flags) })?;
+        Ok(unsafe { Journal::from_ptr(jp.assume_init()) })
+    }
+
+    #[cfg(feature = "systemd_v245")]
     fn open_with_opts_ns<A: CStrArgument>(
         namespace: Option<A>,
         opts: &OpenOptions,

